Create aleth-bootnode executable and remove "mode" argument from Aleth

halfalicious opened the pull request:
Fix #4855, #4703 

Create the aleth-bootnode executable which just starts the `Host `thread in discovery mode with no p2p functionality. This is the same approach that Geth's "bootnode" binary takes (please see #4703 for details).

Configuration options:
```

NAME:
   aleth-bootnode
USAGE:
   aleth-bootnode [options]

GENERAL OPTIONS:
  -h [ --help ]         Show this help message and exit

NETWORKING:
  --public-ip <ip>          Force advertised public IP to the given IP (default: auto)
  --listen-ip <ip>(:<port>) Listen on the given IP for incoming connections (default: 0.0.0.0)
  --listen <port>           Listen on the given port for incoming connections (default: 30303)

LOGGING OPTIONS:
  -v [ --log-verbosity ] <0 - 4>        Set the log verbosity from 0 to 4 (default: 2).
  --log-channels <channel_list>         Space-separated list of the log channels to show (default: show all channels).
  --log-exclude-channels <channel_list> Space-separated list of the log channels to hide.
```

I've also removed the "mode" argument from Aleth's command line since it doesn't provide any purpose.

Note that I still need to verify that the bootnode works correctly - in my tests yesterday (running aleth-bootnode and 2 aleth clients) I didn't see the bootnode or either of the clients send FINDNODE packets so I need to investigate and figure out why.

I also need to add `composeClientVersion `functionality before merging.

chfast commented on libdevcore/Common.h:
```diff
@@ -303,4 +303,16 @@ enum class WithExisting: int
 
 void setDefaultOrCLocale();
+
+unsigned lineWidth();
```
Just make it `static constexpr unsigned c_lineWidth = 160`.

gumb0 commented on CMakeLists.txt:
```diff
@@ -123,6 +123,7 @@ add_subdirectory(libwebthree)
 add_subdirectory(libweb3jsonrpc)
 
 add_subdirectory(aleth)
+add_subdirectory(aleth-bootnode)
```
Maybe this should go under `if (TOOLS)` ?

chfast replied:
Yes.

halfalicious replied:
Thanks for the tip, done!

chfast commented:
@halfalicious it builds for me without a problem. Finding `aleth-buildinfo.h` in VS might be tricky, but you can try finding `aleth-buildinfo` project in VS solution and/or check `build dir/aleth/` directory.

gumb0 commented on aleth-bootnode/main.cpp:
```diff
@@ -0,0 +1,129 @@
+                                       NetworkConfig(publicIP, listenIP, listenPort, upnp);
+    auto netData = contents(getDataDir() / fs::path(NetworkConfigFileName));
+
+    // TODO: Compose client version
```
Not that much important for the bootnode I think (it's used in RLPx handshake, but we'll return no capabilities and it won't go any further)

halfalicious replied:
Got it, that makes things simpler (no need for buildinfo.h). Thanks for the tip!

gumb0 commented on aleth/main.cpp:
```diff
@@ -828,8 +790,7 @@ int main(int argc, char** argv)
     dev::WebThreeDirect web3(WebThreeDirect::composeClientVersion("aleth"), db::databasePath(),
-        snapshotPath, chainParams, withExisting, nodeMode == NodeMode::Full ? caps : set<string>(),
-        netPrefs, &nodesState, testingMode);
+        snapshotPath, chainParams, withExisting, caps, netPrefs, &nodesState, testingMode);
```
`caps` parameter doesn't do anything useful anymore, we could remove it

halfalicious replied:
Confirmed - it's only used in the WebThree constructor to determine if the client should be configured / started and we always do this when running Aleth so it can be removed.

gumb0 commented on aleth/main.cpp:
```diff
@@ -989,15 +950,12 @@ int main(int argc, char** argv)
-        if (networkID != NoNetworkID)
-            c->setNetworkId(networkID);
-    }
+    eth::Client* c = web3.ethereum();
```
I would use a reference here to avoid raw pointers

gumb0 commented on libp2p/NodeTable.cpp:
```diff
@@ -32,6 +33,30 @@ namespace
+    uint8_t type;
+};
+
+DatagramTypeTableEntry datagramTypesTable[] = {
```
```suggestion
DatagramTypeTableEntry const c_datagramTypesTable[] = {
```

gumb0 commented on libp2p/NodeTable.cpp:
```diff
@@ -32,6 +33,30 @@ namespace
+    {"NEIGHBORS", Neighbours::type}
+};
+
+string datagramNameFromType(uint8_t _t)
```
I'd suggest adding `typeName()` method to `RLPXDatagramFace` (and each implementation returning its name) instead of this. Then you won't need to handle unknown packet case and no need to throw.

halfalicious replied:
Good idea, I'll go ahead and do that

gumb0 commented on libp2p/NodeTable.cpp:
```diff
@@ -409,6 +434,7 @@ void NodeTable::onReceived(UDPSocketFace*, bi::udp::endpoint const& _from, bytes
             return;
         }
 
+        cnetdetails << "Received " << datagramNameFromType(packet->packetType()) << " from " << packet->sourceid << "@" << _from.address() << ":" << _from.port();
```
I'd suggest outputting this and other your logs to `LOG(m_logger)` (it's lower verbosity (3 - Debug), but for me Debug verbosity is to log everything that happens, and 4 - Trace is for dumping into log actual binary contents of all data packets, and for tracing VM execution)

Also `m_logger` is more efficient than global `cnetdetails`, because it doesn't need to be thread-safe.

halfalicious commented on aleth-bootnode/main.cpp:
```diff
@@ -0,0 +1,129 @@
+    string const NetworkConfigFileName = ProgramName + "-network.rlp";
+
+    /// Networking params.
+    string listenIP;
```
Got it - I do that if I'm going to be using them in multiple places, but I'll go ahead just declare them right before I use them.

halfalicious commented:
> 
> 
> @halfalicious it builds for me without a problem. Finding `aleth-buildinfo.h` in VS might be tricky, but you can try finding `aleth-buildinfo` project in VS solution and/or check `build dir/aleth/` directory.

I always build from the console so it was pretty easy to find `aleth-buildinfo.h`..the problem was that there was no` aleth-bootnode-buildinfo.h`. It's a moot point now anyway since I don't need to compose the client version (since it's not used when performing discovery, only when establishing p2p connections)

halfalicious commented:
Note quite ready for merging yet - I want to add a log message for outgoing ping packets and take another look at the log output and compare it to Geth.

chfast commented:
> he problem was that there was no` aleth-bootnode-buildinfo.h`

There will be no `aleth-bootnode-buildinfo.h`. The prefix in `aleth-buildinfo.h` is for Aleth project and the same header is used by all the tools.

gumb0 commented on aleth-bootnode/main.cpp:
```diff
@@ -0,0 +1,129 @@
+
+    po::options_description generalOptions("GENERAL OPTIONS", lineWidth());
+    auto addGeneralOption = generalOptions.add_options();
+    addGeneralOption("help,h", "Show this help message and exit\n");
```
Yeah I think it provides only the conversion of `options_description` to a string, but `-h` still need to be handled

gumb0 commented on aleth-bootnode/main.cpp:
```diff
@@ -0,0 +1,150 @@
+    auto netData = contents(getDataDir() / fs::path(c_networkConfigFileName));
+
+    // Pass in empty client version since it's not used in the discovery process
+    Host h("", netPrefs, &netData);
```
Still I think better to give `c_programName` as client name; if someone tries to handshake with us, they'll see it.

halfalicious commented on aleth-bootnode/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+add_executable(aleth-bootnode ${sources})
+target_link_libraries(
+    aleth-bootnode
+    PRIVATE devcore Boost::program_options p2p
```
For future reference, what's the general rule used for ordering the dependencies? Is it from most specialized to least specialized internal libraries followed by 3rd-party libraries?

chfast replied:
> most specialized to least specialized internal libraries followed by 3rd-party libraries?

Exactly. But this is not very important though.

halfalicious commented:
This should be good to go now...I addressed the remaining issues, added ping logging, and cleaned up the change history a bit and removed the merge commits. I also slightly simplified the discovery packet log statements (I didn't have to explicitly log the destination endpoint's port as the << operator is overloaded).

gumb0 commented on libp2p/NodeTable.cpp:
```diff
@@ -299,7 +295,8 @@ void NodeTable::evict(shared_ptr<NodeEntry> _leastSeen, shared_ptr<NodeEntry> _n
     if (evicts == 1)
         doCheckEvictions();
-    ping(_leastSeen.get());
+    if (_leastSeen)
```
I think this `if` is not needed, it's already checked at the calling side, and `_lastSeen` is already used above in this function anyway

chfast replied:
@gumb0 Can you fix it now? I'd like to make a build with aleth-bootnode included.
